crates/typst/src/engine.rs

```diff
@@ -146,20 +186,90 @@ impl Sink {
     pub fn delayed(&mut self) -> EcoVec<SourceDiagnostic> {
         std::mem::take(&mut self.delayed)
     }
 
-    /// Get the stored warnings.
-    pub fn warnings(self) -> EcoVec<SourceDiagnostic> {
-        self.warnings
-    }
-
     /// Get the values for the traced span.
     pub fn values(self) -> EcoVec<(Value, Option<Styles>)> {
         self.values
     }
+
+    /// Apply warning suppression, deduplication, and return the remaining
+    /// warnings.
+    ///
+    /// We deduplicate warnings which are identical modulo tracepoints.
+    /// This is so we can attempt to suppress each tracepoint separately,
+    /// without having one suppression discard all other attempts of raising
+    /// this same warning through a different set of tracepoints.
+    ///
+    /// For example, calling the same function twice but suppressing a warning
+    /// on the first call shouldn't suppress on the second, so each set of
+    /// tracepoints for a particular warning matters. If at least one instance
+    /// of a warning isn't suppressed, the warning will be returned, and the
+    /// remaining duplicates are discarded.
+    pub fn suppress_and_deduplicate_warnings(
+        mut self,
+        world: &dyn World,
+    ) -> EcoVec<SourceDiagnostic> {
+        let mut unsuppressed_warning_set = HashSet::<u128>::default();
+
+        // Only retain warnings which weren't locally suppressed where they
+        // were emitted or at any of their tracepoints.
+        self.warnings.retain(|diag| {
+            let hash =
+                crate::utils::hash128(&(&diag.span, &diag.identifier, &diag.message));
+            if unsuppressed_warning_set.contains(&hash) {
+                // This warning - with the same span, identifier and message -
+                // was already raised and not suppressed before, with a
+                // different set of tracepoints. Therefore, we should not raise
+                // it again, and checking for suppression is unnecessary.
+                return false;
+            }
+
+            let Some(identifier) = &diag.identifier else {
+                // Can't suppress without an identifier.
+                // Therefore, retain the warning.
+                // It is not a duplicate due to the check above.
+                unsuppressed_warning_set.insert(hash);
+                return true;
+            };
+
+            let should_raise = !check_warning_suppressed(diag.span, world, identifier)
+                && !diag.trace.iter().any(|tracepoint| {
+                    check_warning_suppressed(tracepoint.span, world, identifier)
+                });
+
+            // If this warning wasn't suppressed, any further duplicates (with
+            // different tracepoints) should be removed.
+            should_raise && unsuppressed_warning_set.insert(hash)
+        });
+
+        self.warnings
+    }
+
+    /// Adds a tracepoint to all warnings outside the given span.
+    pub fn trace_warnings<F>(
```

Comment:
Not sure whether this method pulls its weight or whether we should just inline its definition at the only call site.